File: backend/src/app/schemas/legal.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


# ── Legal Documents ──────────────────────────────────────────

class LegalDocumentCreate(BaseModel):
    title: str
    doc_type: str  # escritura | pact | sha | bylaws | board_resolution | other
    date: str | None = None  # ISO date string
    summary: str | None = None
    file_url: str | None = None
    file_name: str | None = None
    notes: str | None = None


class LegalDocumentUpdate(BaseModel):
    title: str | None = None
    doc_type: str | None = None
    date: str | None = None
    summary: str | None = None
    file_url: str | None = None
    file_name: str | None = None
    notes: str | None = None


class LegalDocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    doc_type: str
    date: datetime | None
    summary: str | None
    file_url: str | None
    file_name: str | None
    notes: str | None
    created_at: datetime | None = None


# ── Company Legal Info ───────────────────────────────────────

class CompanyLegalCreate(BaseModel):
    legal_name: str | None = None
    cif: str | None = None
    registered_address: str | None = None
    city: str | None = None
    postal_code: str | None = None
    country: str | None = None
    registration_number: str | None = None
    registration_date: str | None = None  # ISO date string
    notary: str | None = None
    protocol_number: str | None = None
    notes: str | None = None


class CompanyLegalUpdate(BaseModel):
    legal_name: str | None = None
    cif: str | None = None
    registered_address: str | None = None
    city: str | None = None
    postal_code: str | None = None
    country: str | None = None
    registration_number: str | None = None
    registration_date: str | None = None
    notary: str | None = None
    protocol_number: str | None = None
    notes: str | None = None


class CompanyLegalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    company_id: str
    legal_name: str | None
    cif: str | None
    registered_address: str | None
    city: str | None
    postal_code: str | None
    country: str | None
    registration_number: str | None
    registration_date: datetime | None
    notary: str | None
    protocol_number: str | None
    notes: str | None
    created_at: datetime | None = None
    updated_at: datetime | None = None

File: frontend/src/lib/api/legal.ts
import { apiFetch } from "@/lib/api";
import type { LegalDocument, CompanyLegal } from "@/types";

// ── Legal Documents ──────────────────────────────────────────

export async function fetchLegalDocuments(
  companyId: string
): Promise<LegalDocument[]> {
  return apiFetch<LegalDocument[]>(`/legal/${companyId}/documents`);
}

export async function createLegalDocument(
  companyId: string,
  data: {
    title: string;
    doc_type: string;
    date?: string;
    summary?: string;
    file_url?: string;
    file_name?: string;
    notes?: string;
  }
): Promise<LegalDocument> {
  return apiFetch<LegalDocument>(`/legal/${companyId}/documents`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateLegalDocument(
  id: string,
  data: Partial<{
    title: string;
    doc_type: string;
    date: string;
    summary: string;
    file_url: string;
    file_name: string;
    notes: string;
  }>
): Promise<LegalDocument> {
  return apiFetch<LegalDocument>(`/legal/documents/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export async function deleteLegalDocument(id: string): Promise<void> {
  await apiFetch<void>(`/legal/documents/${id}`, { method: "DELETE" });
}

// ── Company Legal Info ───────────────────────────────────────

export async function fetchCompanyLegal(
  companyId: string
): Promise<CompanyLegal | null> {
  return apiFetch<CompanyLegal | null>(`/legal/${companyId}/company-info`);
}

export async function upsertCompanyLegal(
  companyId: string,
  data: Partial<{
    legal_name: string;
    cif: string;
    registered_address: string;
    city: string;
    postal_code: string;
    country: string;
    registration_number: string;
    registration_date: string;
    notary: string;
    protocol_number: string;
    notes: string;
  }>
): Promise<CompanyLegal> {
  return apiFetch<CompanyLegal>(`/legal/${companyId}/company-info`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}
